src/pad/src/ICeWall.cpp

```diff
@@ -0,0 +1,1070 @@
+}
+
+void ICeWall::makeFakeSite(const std::string& name, int width, int height)
+{
+  const std::string lib_name = "FAKE_IO";
+  auto* lib = db_->findLib(lib_name.c_str());
+  if (lib == nullptr) {
+    lib = odb::dbLib::create(db_, lib_name.c_str());
+  }
+
+  auto* site = odb::dbSite::create(lib, name.c_str());
+  site->setWidth(width);
+  site->setHeight(height);
+}
+
+void ICeWall::makeIORow(odb::dbSite* horizontal_site,
+                        odb::dbSite* vertical_site,
+                        odb::dbSite* corner_site,
+                        int west_offset,
+                        int north_offset,
+                        int east_offset,
+                        int south_offset,
+                        odb::dbOrientType rotation,
+                        int ring_index)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  if (horizontal_site == nullptr) {
+    horizontal_site = vertical_site;
+  }
+
+  if (horizontal_site == nullptr || vertical_site == nullptr
+      || corner_site == nullptr) {
+    return;
+  }
+
+  std::string name_format;
+  if (ring_index >= 0) {
+    name_format = fmt::format("IO_{{}}_{}", ring_index);
+  } else {
+    name_format = "IO_{}";
+  }
+
+  const odb::Rect die = block->getDieArea();
+
+  odb::Rect outer_io(die.xMin() + west_offset,
+                     die.yMin() + south_offset,
+                     die.xMax() - east_offset,
+                     die.yMax() - north_offset);
+
+  const int cheight = corner_site->getHeight();
+  const int cwidth
+      = std::max(vertical_site->getHeight(), corner_site->getWidth());
+
+  const int x_sites = std::ceil(static_cast<double>(outer_io.dx() - 2 * cwidth)
+                                / vertical_site->getWidth());
+  outer_io.set_xhi(outer_io.xMin() + 2 * cwidth
+                   + x_sites * vertical_site->getWidth());
+  const int y_sites = std::ceil(static_cast<double>(outer_io.dy() - 2 * cheight)
+                                / horizontal_site->getWidth());
+  outer_io.set_yhi(outer_io.yMin() + 2 * cheight
+                   + y_sites * horizontal_site->getWidth());
+
+  const odb::Rect corner_origins(outer_io.xMin(),
+                                 outer_io.yMin(),
+                                 outer_io.xMax() - cwidth,
+                                 outer_io.yMax() - cheight);
+
+  const odb::dbTransform xform(rotation);
+
+  // Create corners
+  const int corner_sites
+      = std::max(horizontal_site->getHeight(), corner_site->getWidth())
+        / corner_site->getWidth();
+  auto create_corner = [block, corner_site, corner_sites, &name_format, &xform](
+                           const std::string& name,
+                           const odb::Point& origin,
+                           odb::dbOrientType orient) -> odb::dbRow* {
+    const std::string row_name = fmt::format(name_format, name);
+    odb::dbTransform rotation(orient);
+    rotation.concat(xform);
+    return odb::dbRow::create(block,
```

Comment:
dbBlock::getCoreArea() returns the bbox of the rows.  This will case the core area to expand out to the pads.  If you are going to use rows we'll need to rethink how the core area gets defined/stored.  The current icewall doesn't create rows.

Reply:
it could be as simple as checking the row type before using it for the core area, but I think PDNgen and DPL will likely need to be updated to handle this correctly.

Reply:
@maliberty I can go through the code to find the places `getRows()` is used and either (create a `getCoreRows()` function which preserves the current functionality and use that instead, or add the check for the site class type in the places where it's being used)? Which do you think makes sense?

Reply:
What do you mean by "row type"?  Do you mean the class of the site used by the row?  I think that would be fine to eliminate class=PAD from the core area calculation.

Reply:
You'll also need to make sure gpl, dpl, and dpo understand to ignore PAD rows.